### PR TITLE
Fix pad_across_processes to allow negative dimensions

### PR DESCRIPTION
**Pull request by harish-kamath**
# What does this PR do?

Fix the `pad_across_processes` function so that you can input negative dimensions. Right now, `indices` will always have slices of `(None, None)` if `dim` is negative.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Could you please run `make quality` and `make style`?

**Comment by harish-kamath** [comment]
@BenjaminBossan I ran it, but seems like there's no file changes. Might be from a different file, I merged

```
accelerate git:(main) make quality
black --required-version 23 --check tests src examples benchmarks utils
All done! ✨ 🍰 ✨
118 files would be left unchanged.
ruff tests src examples benchmarks utils
doc-builder style src/accelerate docs/source --max_len 119 --check_only

accelerate git:(main) make style
black --required-version 23 tests src examples benchmarks utils
All done! ✨ 🍰 ✨
118 files left unchanged.
ruff tests src examples benchmarks utils --fix
doc-builder style src/accelerate docs/source --max_len 119
```

**Comment by BenjaminBossan** [comment]
> but seems like there's no file changes

Strange, not sure what the issue is, maybe something with the installed versions of ruff. Anyway, the lines you touched are too long, you'd have to add a line break to get to 119 chars.

**Comment by harish-kamath** [comment]
Ok should be fixed @BenjaminBossan

File: src/accelerate/utils/operations.py [hunk 1]
```diff
@@ -577,12 +577,13 @@ def _pad_across_processes(tensor, dim=0, pad_index=0, pad_first=False):
         new_size = list(old_size)
         new_size[dim] = max_size
         new_tensor = tensor.new_zeros(tuple(new_size)) + pad_index
+        is_dim = lambda i: i == dim or i == len(new_size) - dim
```
Comment (muellerzr) [inline src/accelerate/utils/operations.py:580]: The issue here is lambda. Just do:
```python
def _is_neg_dim(dim):
    return i == dim or (i == len(new_size) - dim)
```

**Review by muellerzr — COMMENTED** [review]
Thanks! Overall this looks okay, but to fix the quality issue it has issues with using a lambda. Make an inner func instead please :)

**Comment by muellerzr** [comment]
Gentle ping @harish-kamath :) 

(Also try doing `pip install -e .[quality]` to perhaps have `make style; make quality` pass)

**Comment by harish-kamath** [comment]
Sorry just fixed, probably just easier to do it this way

**Comment by harish-kamath** [comment]
Still having trouble with style/quality - `pip install -e .[quality]` results in `no matches found: .[quality]`.

**Comment by muellerzr** [comment]
Make sure you’re running that from the root directory of the accelerate folder

**Comment by harish-kamath** [comment]
Yep I did. Interestingly, I think it just doesn't interpret it as a pip command? `no command found` seems to actually think that the `.[quality]` is part of another command. `pip install -e .` works as expected though.

**Comment by harish-kamath** [comment]
Note for anyone else that may have trouble installing with pip and square brackets:
If you are using a zsh shell, you should use '' around your command e.g. `pip install -e '.[quality]'`

**Comment by harish-kamath** [comment]
That being said - there is still no issue locally with style or quality having done this @muellerzr and the actual PR is really small, so not sure what is failing

**Comment by muellerzr** [comment]
Try rebasing from main, iirc this was an issue with deepspeed back then

**Pull request closed without merging**